Feature/channel summation

psurukuc opened the pull request:
Not fully ready to merge. Need to confirm comments and such.

psurukuc commented:
Andrew has been implementing his time-dependent grid in Katydid and that depends on some of the updates on this branch. Plan to merge this branch to develop and then push those changes into feature/time-domain-ch-sum branch before requesting a PR.

nsoblath commented on Source/SpectrumAnalysis/KTChannelAggregator.cc:
```diff
@@ -17,15 +19,22 @@ namespace Katydid
+        fSummedFrequencyData("agg-fft", this),
+        fPhaseChFrequencySumSlot("fft", this, &KTChannelAggregator::SumChannelVoltageWithPhase, &fSummedFrequencyData),
+        fAxialSumSlot("ax-agg-fft", this, &KTChannelAggregator::SumChannelVoltageWithPhase, &fSummedFrequencyData),
+        fActiveRadius(0.0516),
```
Values should be initialized to things that are valid values for those variables, but will generally _not_ be useful in an analysis.  This way all users are made to explicitly specify things in their configs and not rely on default values that might get changed. So, e.g. numbers of things initialized to 0; maxes and mins to 0 and 1, respectively; wavelength to 0.

nsoblath commented on Source/SpectrumAnalysis/KTChannelAggregator.cc:
```diff
@@ -75,104 +101,199 @@ namespace Katydid
-	    fAntiSpiralPhaseShifts.insert(channelPhaseShift);
-	}
-	    return true;
+        for(int i=0;i<channelCount;++i)
```
`int` --> `unsigned` and spaces between arguments

nsoblath commented on Source/SpectrumAnalysis/KTChannelAggregator.cc:
```diff
@@ -75,104 +101,199 @@ namespace Katydid
 
     bool KTChannelAggregator::SumChannelVoltageWithPhase(KTFrequencySpectrumDataFFTW& fftwData)
+    {
+        KTAggregatedFrequencySpectrumDataFFTW& newAggFreqData = fftwData.Of< KTAggregatedFrequencySpectrumDataFFTW >().SetNComponents(0);
```
`int` --> `unsigned` and spaces between arguments

nsoblath commented on Source/SpectrumAnalysis/KTChannelAggregator.cc:
```diff
@@ -75,104 +101,199 @@ namespace Katydid
         // Get the number of frequency bins from the first component of fftwData
         int nFreqBins = freqSpectrum->GetNFrequencyBins();
-        int nComponents = fftwData.GetNComponents(); // Get number of components
+        int nTotalComponents = fftwData.GetNComponents(); // Get number of components
```
`int` --> `unsigned` and more spaces.
Either `if (true)` or `if( true )`, but be consistent.

nsoblath commented on Source/SpectrumAnalysis/KTChannelAggregator.hh:
```diff
@@ -71,21 +78,53 @@ namespace Katydid
+            MEMBERVARIABLE(bool, IsPartialRing);
+
+            // If IsPartialRing is true, the total number of channels is the product of the number of channels extracted from the egg file and the PartialRingMultiplicity
+            MEMBERVARIABLE(int, PartialRingMultiplicity)
```
unsigned?

nsoblath commented on Source/SpectrumAnalysis/KTChannelAggregator.hh:
```diff
@@ -71,21 +78,53 @@ namespace Katydid
+            MEMBERVARIABLE(double, SummationMaxFreq);
+
+            // Number of axial rings/subarrays
+            MEMBERVARIABLE(int, NRings);
```
unsigned

nsoblath commented on Source/SpectrumAnalysis/KTChannelAggregator.hh:
```diff
@@ -71,21 +78,53 @@ namespace Katydid
+             * If fIsUserDefinedGrid is false, defines a square grid with fNGrid*fNGrid points
+             * Returns total number of grid points defined
+             */
+            int DefineGrid(KTAggregatedFrequencySpectrumDataFFTW &newAggFreqData);
```
return value `unsigned`?

nsoblath commented on Source/SpectrumAnalysis/KTChannelAggregator.hh:
```diff
@@ -12,6 +12,7 @@
 #include "KTData.hh"
```
This file has a bunch of tab characters in it.  Please convert all to 4 spaces

nsoblath reviewed (changes requested):
Very nice overall.  I didn't check all of the math, or anything, but for the most part the changes looked consistent with best practices and standards, both for C++ in general and Katydid specifically.  

Is there a test program for the channel aggregators?  If not, please add one that verifies that the processors are behaving as expected.

I have some comments that apply in many places, and I've tried to include individual comments where I've noticed individual instances, but some may have been missed.

* There are a number of places where spaces are missing between function arguments (calls and definitions) or where multiple variables are defined in one line.  Please add those spaces.
* Make sure to initialize variables where they're defined to avoid the possibility that they're used uninitialized.
* Standard practice in Katydid is to use unsigned ints for counting variables (like components or bins) if there's no chance of using a negative number.  In some cases I was pretty sure unsigned variables should be used.  In some cases I wasn't sure whether or not there was a possibility of a negative value.  Please check these and use unsigned in any case where negative values aren't expected.
* There were a few hard tab characters in at least one file.

psurukuc commented on Examples/ConfigFiles/GridPoints.txt:
```diff
@@ -0,0 +1,8 @@
+# Template for grid point file
```
Hi Noah, yes, there is no compelling reason to use text file. I work on chaging this to YAML file. Just so I unserstand correctly, you want that to be done in a new feature branch right after merging this one into develop ?

nsoblath replied:
That's right.  As I mentioned in one of the other comments, let's rename the relevant config parameters so they don't say "text" in their names, in anticipation of the change.  But let's not hold up the merge for this particular change.

psurukuc commented on Source/IO/Conversions/KT2ROOT.cc:
```diff
@@ -477,44 +478,63 @@ namespace Katydid
+        std::vector<unsigned> pointIndex;
+        pointIndex.assign(nZ,0);
+
+        for(int iZ=0;iZ<nZ;iZ++)
```
Done

psurukuc commented on Source/IO/Conversions/KT2ROOT.cc:
```diff
@@ -477,44 +478,63 @@ namespace Katydid
-            int yBin = iComponents % nGridPoints;
-            hist->SetBinContent(xBin + 1, yBin + 1, aggfs.GetSummedGridVoltage(iComponents));
+            double xPos,yPos,zPos;
+            aggfs.GetGridPoint(iComponents,xPos,yPos,zPos);
```
Done

psurukuc commented on Source/IO/BasicROOTFileWriter/KTBasicROOTTypeWriterSpectrumAnalysis.cc:
```diff
@@ -808,14 +874,17 @@ namespace Katydid
+        conv << "graphAggGridFFTW_" << sliceNumber;
+        string graphName;
+        conv >> graphName;
+        std::vector<TGraph2D*> aggregatedGridGraphs = KT2ROOT::CreateGridGraphs(sumData,graphName);
```
Done

psurukuc commented on Source/SpectrumAnalysis/KTAxialAggregator.cc:
```diff
@@ -0,0 +1,79 @@
+    {
+        if (node != NULL)
+        {
+            fNRings= node->get_value< signed int >("n-rings", fNRings);
```
That is a mistake, thanks for catching.

psurukuc commented:
Hi Noah, thanks for very carefully going through the commits. I think I fixed all the suggestions expect for replacing YAML file with text file.

psurukuc commented:
Actually, please hold off on reviewing. I think I may have to make a few more changes. Also, I realized I haven't actually pushed all my changes to github yet.

psurukuc commented:
@nsoblath, thanks for the comments. I have implemented all the changes. There is a test program called TestChannelAggregator to test the ChannelAggregator and I modified it slightly to fit the new changes better.

nsoblath reviewed (approved):
This all looks good.  Thanks for all of the efforts on the this PR!
